api/python/t4/formats.py

```diff
@@ -0,0 +1,418 @@
+
+        If **kwargs is given, it overrides metadata kwargs, if present.
+
+        Args:
+            bytes_obj: bytes to deserialize
+            meta: object metadata, may contain deserialization prefs
+            **kwargs: metadata to (potentially) use.
+        """
+        # deactivated if this method is overridden.
+        if type(self).deserialize == Format.deserialize:
+            meta = {} if meta is None else meta
+            if getattr(self, '_deserializer', None) is None:
+                raise NotImplementedError()
+            format_meta = meta.get('format', {})
+            deserialization_kwargs = format_meta.get('deserialization', {})
+            deserialization_kwargs.update(kwargs)
+            return self._deserializer(bytes_obj, **deserialization_kwargs)
+
+
+Format('bytes',
+        serializer=lambda obj: obj,
+        deserializer=lambda bytes_obj: bytes_obj,
+        handled_extensions=['bin'],
+        handled_types=[bytes],
+).register()
+
+
+Format('json',
+       serializer=lambda obj, **kwargs: json.dumps(obj, **kwargs).encode('utf-8'),
+       deserializer=lambda bytes_obj, **kwargs: json.loads(bytes_obj.decode('utf-8'), **kwargs),
+       handled_extensions=['json'],
+       handled_types=[dict, list, int, float, str, tuple, type(None)]
+).register()
+
+
+Format('unicode',  # utf-8 instead?
+       serializer=lambda s: s.encode('utf-8'),
+       deserializer=lambda b: b.decode('utf-8'),
+       handled_extensions=['txt', 'md', 'rst'],
+       handled_types=[text_type],
+).register()
+
+
+class NumpyFormat(Format):
+    def __init__(self, name, handled_extensions=None, **kwargs):
+        handled_extensions = [] if handled_extensions is None else handled_extensions
+
+        if 'npy' not in handled_extensions:
+            handled_extensions.append('npy')
+        if 'npz' not in handled_extensions:
+            handled_extensions.append('npz')
+
+        super().__init__(name, handled_extensions=handled_extensions, **kwargs)
+
+    def handles_obj(self, obj):
+        # If this is a numpy object, numpy must be loaded.
+        if 'numpy' not in sys.modules:
+            return False
+        import numpy as np
+        if np.ndarray not in self._handled_types:
+            self._handled_types.append(np.ndarray)
+        return super().handles_obj(obj)
+
+    def _deserializer(self, bytes_obj):
+        import numpy as np
+
+        buf = io.BytesIO(bytes_obj)
+        return np.load(buf, allow_pickle=False)
+
+    def _serializer(self, obj):
+        import numpy as np
+        buf = io.BytesIO()
+        np.save(buf, obj, allow_pickle=False)
+        return buf.getvalue()
+
+
+NumpyFormat('numpy').register()
+
+
+class ParquetFormat(Format):
+    def __init__(self, name, handled_extensions=None, **kwargs):
+        handled_extensions = [] if handled_extensions is None else handled_extensions
+
+        if 'parquet' not in handled_extensions:
+            handled_extensions.append('parquet')
```

Comment:
Probably not good to modify caller's data... but really, I'd just hardcode the extensions for parquet and numpy for now.